### ISSUE TITLE
etcd keeps getting started in place of etcd2

### ISSUE THREAD

**Issue by hookenz** [coreos/etcd#3211]
I have one etcd2 cluster node that keeps starting etcd instead of etcd2 on reboot.
This causes all sorts of instability to the cluster such as units randomly failing and attempting to restart on a continual basis until they finally do.  The cycle repeats to other units scheduled in the cluster.

My cloud config in /usr/share/oem/cloud-config.yml is told to start etcd2 not etcd.  I've confirmed this many times.   However, before etc2 even gets a chance to start, etcd is already running.
There is nothing starting it in /etc/systemd/system either.

It also maybe highlighting a bug.  An etcd server is transitioning to master when joining an etcd2 cluster that it's meant to know nothing about.  It even has a different key.  Or should do.

I suspect this is some sort of hangover from transitioning to etcd2.  But I fail to see why it's being started at all.

systemctl status etcd shows:

```
● etcd.service - etcd
   Loaded: loaded (/usr/lib64/systemd/system/etcd.service; static; vendor preset: disabled)
   Active: failed (Result: exit-code) since Mon 2015-08-03 03:51:01 UTC; 16min ago
 Main PID: 974 (code=exited, status=2)

Aug 03 03:51:01 service4-1 etcd[974]: [etcd] Aug  3 03:51:01.910 INFO      | 9347e5d333af492caf1fb4835af1f6ee starting in peer mode
Aug 03 03:51:01 service4-1 etcd[974]: [etcd] Aug  3 03:51:01.910 INFO      | 9347e5d333af492caf1fb4835af1f6ee: state changed from 'initialized' to 'follower'.
Aug 03 03:51:01 service4-1 etcd[974]: [etcd] Aug  3 03:51:01.910 INFO      | 9347e5d333af492caf1fb4835af1f6ee: state changed from 'follower' to 'leader'.
Aug 03 03:51:01 service4-1 etcd[974]: [etcd] Aug  3 03:51:01.910 INFO      | 9347e5d333af492caf1fb4835af1f6ee: leader changed from '' to '9347e5d333af492caf1fb4835af1f6ee'.
Aug 03 03:51:01 service4-1 systemd[1]: Stopping etcd...
Aug 03 03:51:01 service4-1 systemd[1]: etcd.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
Aug 03 03:51:01 service4-1 systemd[1]: Stopped etcd.
Aug 03 03:51:01 service4-1 systemd[1]: etcd.service: Unit entered failed state.
Aug 03 03:51:01 service4-1 systemd[1]: etcd.service: Failed with result 'exit-code'.
Aug 03 03:51:33 service4-1 systemd[1]: Stopped etcd.
```

No other nodes are doing this.  How can I diagnose what could be starting etcd?

**Comment by carmstrong** [comment]
@hookenz You can try masking etcd with this in your cloud-config:

```
  - name: etcd.service
    mask: true
```

**Comment by hookenz** [comment]
@camstrong - Good suggestion, I'll try that.

**Comment by hookenz** [comment]
@camstrong - That helped, at least now etcd2 starts.  It just means I have a failed unit.  The result:

```
● etcd.service
   Loaded: masked (/dev/null)
   Active: failed (Result: exit-code) since Mon 2015-08-03 04:36:15 UTC; 2min 30s ago
 Main PID: 974 (code=exited, status=2)
```

**Comment by carmstrong** [comment]
FWIW, we saw this too. The suspicion was that there was some systemd dependency that explicitly listed etcd - maybe fleet? We worked around it by masking etcd as I showed you here, and not caring that the service failed.

**Comment by hookenz** [comment]
@carmstrong - would still like to get to the bottom of it.  I wondered if it was fleet but a grep showed up nothing.  I even checked etcd keys.  I used to run a unit called etcd in place of the default i.e. /etc/systemd/system/etcd.service.  It was a the containerised version of fleet.  But this has been removed, and the install link as well.  So I'm baffled by it.

**Comment by carmstrong** [comment]
> would still like to get to the bottom of it.

Oh, me too! I was just giving it a :+1: as confirmation that it wasn't isolated to you.

> So I'm baffled by it.

Me too.

**Comment by crawford** [comment]
Can you try running `systemctl list-dependencies etcd.service --reverse`? This will show you everything that depends on etcd.service.

**Comment by carmstrong** [comment]
@crawford On 647.2.0 at least, fleet wants etcd:
```console
core@deistest18-coreos-0 ~ $ systemctl list-dependencies etcd.service --reverse
etcd.service
● └─fleet.service
core@deistest18-coreos-0 ~ $ systemctl cat fleet.service
# /usr/lib64/systemd/system/fleet.service
[Unit]
Description=fleet daemon

Wants=etcd.service
After=etcd.service

Wants=fleet.socket
After=fleet.socket

[Service]
ExecStart=/usr/bin/fleetd
Restart=always
RestartSec=10s

# /run/systemd/system/fleet.service.d/20-cloudinit.conf
[Service]
Environment="FLEET_ETCD_REQUEST_TIMEOUT=3"
Environment="FLEET_PUBLIC_IP=100.115.66.71"
```

**Comment by crawford** [comment]
@carmstrong ah, in 653.0.0, we removed the dependency.

**Comment by carmstrong** [comment]
> ah, in 653.0.0, we removed the dependency.

We're waiting on Docker 1.7.1 to land in stable because 1.6.x is fundamentally broken :(

**Comment by hookenz** [comment]
```
core@service4-1 /usr/lib64/systemd/system $ systemctl list-dependencies etcd.service --reverse
etcd.service
```

**Comment by crawford** [comment]
@hookenz hmm, so something must be manually starting etcd. Can you check `journalctl --boot --no-pager`? That will show you roughly when it was started. Maybe it is a service started by coreos-cloudinit, or coreos-cloudinit itself.

**Comment by hookenz** [comment]
It appears to be a dropin unit.  As at one time I was starting it from cloud-init.


```
-- Logs begin at Sat 2015-08-01 18:34:48 UTC. --
Aug 05 21:36:39 service4-1 bash[882]: 2015/08/05 21:36:39 Wrote drop-in unit "20-cloudinit.conf"
Aug 05 21:36:39 service4-1 bash[882]: 2015/08/05 21:36:39 Ensuring runtime unit file "fleet.service" is unmasked
Aug 05 21:36:39 service4-1 bash[882]: 2015/08/05 21:36:39 Ensuring runtime unit file "locksmithd.service" is unmasked
Aug 05 21:36:39 service4-1 bash[882]: 2015/08/05 21:36:39 Calling unit command "restart" on "docker.service"'
Aug 05 21:36:39 service4-1 bash[882]: 2015/08/05 21:36:39 Result of "restart" on "docker.service": done
Aug 05 21:36:39 service4-1 bash[882]: 2015/08/05 21:36:39 Calling unit command "start" on "etcd.service"'
Aug 05 21:36:39 service4-1 bash[882]: Failed to apply cloud-config: Unit etcd.service is masked.
```

**Comment by crawford** [comment]
It looks like your cloud-init is explicitly starting etcd.service.

**Comment by hookenz** [comment]
That's what it looks like.  But can you see anywhere in my cloud-init where it's doing that?  here is my cloud-init.

```
#!/bin/sh
export NAME=service4-1

get_ipv4() {
    IFACE="${1}"
    PREFIX="${2}"
    local ip
    while [ -z "${ip}" ]; do
        ip=$(ip -4 -o addr show dev "${IFACE}" scope global |  grep "inet $PREFIX" | gawk 'NR==1 {split ($4, out, "/"); print out[1]}')
        sleep .1
    done

    echo "${ip}"
}

export COREOS_PRIVATE_IPV4=$(get_ipv4 ib0 "10.10.10")
export COREOS_PUBLIC_IPV4=$COREOS_PRIVATE_IPV4

workdir=$(mktemp --directory)
trap "rm --force --recursive ${workdir}" SIGINT SIGTERM EXIT


cat > "${workdir}/cloud-config.yml" <<EOF
#cloud-config

hostname: ${NAME}
ssh_authorized_keys:
  - ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDlspbYZcWSUm5TYEawKdLpwVDKRo92AcrAiNcvl/5bNzMh+WTHbTgwDDeJGXfRpOtWKjcTHul+FtwyMbK1CLxhbq5RGGq1uYJA+sTIEMQ3/toD2mF8dkQszBDMPRj7RGY5HJ22yd8ozayFlFCGd/9je9KLj8m9YXzZfAoowzTGn0sYpD1/clOTxQQTyUtNlHnYYO8blJrkOncI0lU7XJ+Nhf5UQin1nXVRlsxG96WpmZuqewtqNqYXsx5HEM4hXqOFsslLC5vw5dNywRR9Io414s+N0HLfeffC7+BxEwYaP5rcgKtaGnnnU+EDBlcUKFoM/q19PWQ0TjMrGODcWC1r
write_files:
  - path: /root/.dockercfg
    owner: root
    permissions: 0644
    content: |
      {
          "registry.otoyserver.com": {
                "auth": "bm9kZTpwMlBVWk5BUA==",
                "email": ""
          }
      }
  - path: /etc/systemd/system/docker.service.d/increase-ulimit.conf
    owner: root
    permissions: 0600
    content: |
      [Service]
      LimitMEMLOCK=infinity
      LimitNProc=infinity
  - path: /etc/sysctl.d/10-disable-ipv6.conf
    permissions: 0644
    owner: root
    content: |
      net.ipv6.conf.ib0.disable_ipv6 = 1
coreos:
  etcd2:
    name: ${NAME}.cluster
    discovery-srv: cluster
    initial-cluster-token: cluster1
    initial-cluster-state: new
    initial-advertise-peer-urls: http://${COREOS_PRIVATE_IPV4}:2380
    advertise-client-urls: http://${COREOS_PRIVATE_IPV4}:2379,http://${COREOS_PRIVATE_IPV4}:4001
    listen-client-urls: http://0.0.0.0:2379,http://0.0.0.0:4001
    listen-peer-urls: http://0.0.0.0:2380
  fleet:
    public-ip: ${COREOS_PRIVATE_IPV4}
    metadata: role=services
  units:
    - name: etcd.service
      mask: true
    - name: etcd2.service
      command: start
    - name: fleet.service
      command: start
EOF

coreos-cloudinit --from-file="${workdir}/cloud-config.yml"
```

**Comment by hookenz** [comment]
That's what it looks like.  But I can't see it starting from cloud-init.  Here is my /usr/share/oem/cloud-config.yml file:

That's what it looks like.  But can you see anywhere in my cloud-init where it's doing that?  here is my cloud-init.

```
#!/bin/sh
export NAME=service4-1

get_ipv4() {
    IFACE="${1}"
    PREFIX="${2}"
    local ip
    while [ -z "${ip}" ]; do
        ip=$(ip -4 -o addr show dev "${IFACE}" scope global |  grep "inet $PREFIX" | gawk 'NR==1 {split ($4, out, "/"); print out[1]}')
        sleep .1
    done

    echo "${ip}"
}

export COREOS_PRIVATE_IPV4=$(get_ipv4 ib0 "10.10.10")
export COREOS_PUBLIC_IPV4=$COREOS_PRIVATE_IPV4

workdir=$(mktemp --directory)
trap "rm --force --recursive ${workdir}" SIGINT SIGTERM EXIT


cat > "${workdir}/cloud-config.yml" <<EOF
#cloud-config

hostname: ${NAME}
ssh_authorized_keys:
  - ssh-rsa *** rsa-key here ***
write_files:
  - path: /root/.dockercfg
    owner: root
    permissions: 0644
    content: |
      {
          "registry.otoyserver.com": {
                "auth": "bm9kZTpwMlBVWk5BUA==",
                "email": ""
          }
      }
  - path: /etc/systemd/system/docker.service.d/increase-ulimit.conf
    owner: root
    permissions: 0600
    content: |
      [Service]
      LimitMEMLOCK=infinity
      LimitNProc=infinity
  - path: /etc/sysctl.d/10-disable-ipv6.conf
    permissions: 0644
    owner: root
    content: |
      net.ipv6.conf.ib0.disable_ipv6 = 1
coreos:
  etcd2:
    name: ${NAME}.cluster
    discovery-srv: cluster
    initial-cluster-token: cluster1
    initial-cluster-state: new
    initial-advertise-peer-urls: http://${COREOS_PRIVATE_IPV4}:2380
    advertise-client-urls: http://${COREOS_PRIVATE_IPV4}:2379,http://${COREOS_PRIVATE_IPV4}:4001
    listen-client-urls: http://0.0.0.0:2379,http://0.0.0.0:4001
    listen-peer-urls: http://0.0.0.0:2380
  fleet:
    public-ip: ${COREOS_PRIVATE_IPV4}
    metadata: role=services
  units:
    - name: etcd.service
      mask: true
    - name: etcd2.service
      command: start
    - name: fleet.service
      command: start
EOF

coreos-cloudinit --from-file="${workdir}/cloud-config.yml"
```

**Comment by crawford** [comment]
That looks fine. What provider are you using? Do you have any other cloud-configs floating around?

**Comment by hookenz** [comment]
Bare metal.  There is only one yml file in /usr/share/oem.

Where else can I look?

**Comment by crawford** [comment]
/var/lib/coreos-install is the first thing that comes to mind.

**Comment by hookenz** [comment]
Yes that's it.  There is a file called user_data in there that contains a command to start etcd.
Should I delete that?  Is this file created when you first install coreos and provide a config file?

It seems that providing a new one in /usr/share/oem doesn't mask that out entirely.  I didn't realise this.

**Comment by crawford** [comment]
Woo! That is created by `coreos-install -c blah`, so it's safe to consolidate the two files.

Yeah, there about a million different ways coreos-cloudinit can start. We are fixing this in [Ignition](http://github.com/coreos/ignition).

**Comment by hookenz** [comment]
Yes, looking forward to trying ignition when it comes into the stable release.  cloud-init is really problematic.

**Comment by hookenz** [comment]
Yay, that fixed it!  there was nothing special in the user_data file.  Just the ssh pub key and the starting of etcd.

**Comment by crawford** [comment]
Excellent.

Ignition is in the Alpha channel if you want to take a peak. It should make it into the Stable release in about 3 weeks.

**Comment by hookenz** [comment]
@carmstrong - check /var/lib/coreos-install/user_data removing that fixed it for me.

**Comment by jessegonzalez** [comment]
I had something similar happen to me, however due to a fault of my own.
I had fleetctl submit'ed a unit file where I had overlooked there being an 'After=etcd.service' instead of 'After=etcd2.service'.

To resolve I performed the following steps:
* sudo systemctl stop fleet.service fleet.socket etcd
* sudo systemctl start etcd2
* sudo systemctl reset-failed
* etcdctl rm/rmdir everything under and including /_coreos.com/fleetc/job/my_failed_unit.service
* sudo systemctl start fleet

Hope this helps others.
